### PR TITLE
add/delete only changed overlap records

### PR DESCRIPTION
**Pull request by jsteverman**
* moves ETAS database ops to Module ETAS

### CODE REVIEW

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:36]: `TABLENAME` not defined here

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:41]: Consider extracting this block to a separate method to better expose what it is we're doing here and appease rubocop

File: bin/etas/update_overlap_table.rb
Comment (aelkiss) [inline bin/etas/update_overlap_table.rb:30]: I'm not sure this is really the right name. This table existed before ETAS and is used for other (albeit less visible and perhaps in some cases theoretical) things like section 108 access and direct SSD access (for Michigan anyway) and for driving 'held by my library' facets in solr.

Prior to ETAS, we generally referred to this table as the "production holdings table". I think calling the "overlap table" is more specific while still being accurate. I'm not immediately sure where to put it -- it has some affinity with reports but isn't a report as such. That may be an argument for leaving it where it is (i.e. not in a submodule) for the time being esp. as ultimately we want to see if we can do away with it and have an API direct to the source instead.

Comment (aelkiss) [reply bin/etas/update_overlap_table.rb:30]: Perhaps the right concept is something like "production holdings" or "relational database holdings", but nothing immediate comes to mind as far as a better choice for the module name here.

File: lib/etas/cluster_update.rb
Comment (aelkiss) [inline lib/etas/cluster_update.rb:37]: This seems a little bit clunky; I'll see if I can propose an alternative.

**Review by aelkiss — CHANGES_REQUESTED** [review]
I think the split into `OverlapTableUpdate` and `ClusterUpdate` makes sense. I think we should try to come up with a better name for the module they're in (admittedly not my forte).

I'll separately comment with thoughts on a test to add to make sure we're doing what we think we are regarding minimizing churn.

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:16]: This should be a parameter to initialize, rather than a constant.

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:31]: I would rather see (in `services.rb`):

```ruby
Services.register(:holdings_db_overlap_table) { Services.holdings_db[:holdings_htitem_htmember] }
```

or something along those lines, and getting rid of the `stub_const` calls

File: spec/etas/cluster_update_spec.rb
Comment (aelkiss) [inline spec/etas/cluster_update_spec.rb:16]: See comment above; I'd really like to get rid of these constants.

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:44]: `loggin` -- typo?

File: lib/etas/overlap_table_update.rb
Comment (aelkiss) [inline lib/etas/overlap_table_update.rb:41]: I'd like to see something like

```ruby
logger.debug("Processing overlap for cluster #{cluster._id}")
```

File: lib/etas/cluster_update.rb
Comment (aelkiss) [reply lib/etas/cluster_update.rb:37]: This seems like it works:

```ruby
    def existing_overlaps
      @existing_overlaps ||= cluster.ht_items.map do |ht_item|
        @overlap_table.filter(volume_id: ht_item.item_id).map(&:to_hash)
      end.flatten
    end
```

If you wanted, you could extract the block body as a separate function, e.g.

```ruby
    def existing_overlaps
      @existing_overlaps ||= cluster.ht_items.map { |h| existing_item_overlap(h) }.flatten
    end

    private
   
    def existing_item_overlap(ht_item)
      @overlap_table.filter(volume_id: ht_item.item_id).map(&:to_hash)
    end
```

**Comment by aelkiss** [comment]
Regarding testing that minimizing churn is working as expected:

The specs are a little bit hard to follow but I think are basically already doing this by checking that the delete count is smaller than the old overlap count & likewise that the add count is smaller than the new overlap count.

You could consider adding a more explicit test like:

```ruby
describe "#deletes" do
  it "does not include things in both the old & new overlaps" do
    do_some_setup
    expect(deletes).not_to contain(thing_that_shouldnt_be_deleted)
  end
end

describe "#adds" do
  it "does not include things in both the old & new overlaps" do
    do_some_setup
    expect(adds).not_to contain(thing_that_shouldnt_need_to_be_re_added)
  end
end
```

Regarding testing that it's actually reducing churn in mysql, probably the way to do that would be to stub the overlap table and then check that it receives the expected number of insert & delete calls:

```ruby
it "doesn't churn" do
  mock_overlap_table = stub(:overlap_table)
  allow(mock_overlap_table).to receive(:filter).and_return(mock_overlap_table)
  expect(mock_overlap_table).to receive(:delete).exactly(M).times
  expect(mock_overlap_table).to receive(:insert).exactly(N).times  

  do_some_setup
  described_class.new(mock_overlap_table, cluster).upsert
end
```

You could I suppose set an expectation that `filter` or `insert` receives the parameters you expect it to; that's probably overkill since we're already checking the actual contents of the table. Doing it this way just allows us to make sure that we aren't calling `insert` or `delete` more than we expected.

File: lib/etas/cluster_update.rb
Comment (jsteverman) [reply lib/etas/cluster_update.rb:37]: I had something similar that was leaving me with nil and nil related errors which is how I ended up with my verbose clunker. Confirmed, yours works.

File: lib/etas/overlap_table_update.rb
Comment (jsteverman) [reply lib/etas/overlap_table_update.rb:31]: I registered it as :relational_overlap_table. I am starting to think references to 'holdings_db' are a problem that needs addressing and don't want to perpetuate it.

File: bin/etas/update_overlap_table.rb
Comment (jsteverman) [reply bin/etas/update_overlap_table.rb]: I gave up on giving these a module name and moved back into lib.

File: lib/etas/overlap_table_update.rb
Comment (jsteverman) [reply lib/etas/overlap_table_update.rb]: Done

File: spec/etas/cluster_update_spec.rb
Comment (jsteverman) [reply spec/etas/cluster_update_spec.rb]: Constants killed.

**Review by aelkiss — APPROVED** [review]
Changes look good to me. Should be good to go once you resolve the merge conflict & squash.

File: spec/cluster_update_spec.rb
Comment (mwarin) [inline spec/cluster_update_spec.rb:116]: maybe don't commit these commented lines?

**Pull request merged**
